File: pkg/@eserstack/posts/application/post-service.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

/**
 * PostService — inbound port interface + multi-platform application service.
 * Accepts a collection of platform connections and routes write operations to
 * the correct SocialApi. Reads (timeline) go through FeedAggregator.
 *
 * All methods return Promise<Result<T, Error>> for @eserstack/functions compatibility.
 * Internally uses run() + yield* do-notation for composing sequential operations.
 */

import * as results from "@eserstack/primitives/results";
import * as fn from "@eserstack/functions";
import type { Post } from "../domain/entities/post.ts";
import type { UsageData } from "../domain/entities/usage.ts";
import type { Platform } from "../domain/values/platform.ts";
import type { PostId } from "../domain/values/post-id.ts";
import type { FeedAggregator, PlatformConnection } from "./feed-aggregator.ts";
import type { Scheduler } from "./scheduler.ts";
import type { SocialApi } from "./social-api.ts";
import type { Translator } from "./translator.ts";

/** Per-platform result from a cross-platform operation. */
export interface PostResult {
  platform: Platform;
  post?: Post;
  error?: Error;
}

/** Inbound port: public operations exposed to UI adapters (TUI, web, CLI). */
export interface InboundPostService {
  /** Publish a post immediately. Defaults to the first configured platform. */
  composePost(
    text: string,
    platform?: Platform,
  ): Promise<results.Result<Post, Error>>;
  /** Cross-post the same text to all configured platforms. Returns one entry per platform. */
  composePostToAll(
    text: string,
  ): Promise<results.Result<PostResult[], Error>>;
  /** Translate text, then publish. */
  translateAndPost(params: {
    text: string;
    from: string;
    to: string;
    platform?: Platform;
  }): Promise<results.Result<Post, Error>>;
  /** Queue a post for delivery at a future time. */
  schedulePost(
    params: { text: string; scheduledAt: Date; platform?: Platform },
  ): Promise<results.Result<void, Error>>;
  /** Get unified timeline from all authenticated platforms, newest first. */
  getUnifiedTimeline(
    params?: { maxResultsPerPlatform?: number },
  ): Promise<results.Result<Post[], Error>>;
  /** Get timeline from a specific (or default) platform. */
  getTimeline(
    params?: { maxResults?: number; platform?: Platform },
  ): Promise<results.Result<Post[], Error>>;
  /** Fetch a single post by ID from the given platform. */
  getPost(
    id: PostId,
    platform: Platform,
  ): Promise<results.Result<Post, Error>>;
  /** Post a reply. The platform is inferred from inReplyToPost.platform. */
  replyToPost(params: {
    text: string;
    inReplyToPost: Post;
  }): Promise<results.Result<Post, Error>>;
  /**
   * Post a thread to the given platform.
   * Returns Fail with ThreadPartialError on partial failure.
   */
  postThread(
    texts: ReadonlyArray<string>,
    platform?: Platform,
  ): Promise<results.Result<Post[], Error>>;
  /** Translate text, then post as a reply. Platform inferred from inReplyToPost. */
  translateAndReply(params: {
    text: string;
    inReplyToPost: Post;
    from: string;
    to: string;
  }): Promise<results.Result<Post, Error>>;
  /** Retrieve API usage data for a platform. */
  getUsage(platform?: Platform): Promise<results.Result<UsageData, Error>>;

  /** Repost (share) a post on the given platform without commentary. */
  repost(params: {
    id: PostId;
    platform: Platform;
  }): Promise<results.Result<void, Error>>;
  /** Remove a repost on the given platform. */
  undoRepost(params: {
    id: PostId;
    platform: Platform;
  }): Promise<results.Result<void, Error>>;
  /**
   * Quote post — share a post with your own commentary on the given platform.
   * Returns the newly created Post including the referencedPosts entry.
   */
  quotePost(
    params: { text: string; quotedPostId: PostId; platform: Platform },
  ): Promise<results.Result<Post, Error>>;

  /** Search for posts on a specific platform. */
  searchPosts(
    params: { query: string; maxResults?: number; platform: Platform },
  ): Promise<results.Result<Post[], Error>>;
  /** Search across all authenticated platforms, merge results sorted newest first. */
  searchPostsAll(
    params: { query: string; maxResultsPerPlatform?: number },
  ): Promise<results.Result<Post[], Error>>;
  /** Add a post to the authenticated user's bookmarks. */
  bookmarkPost(params: {
    id: PostId;
    platform: Platform;
  }): Promise<results.Result<void, Error>>;
  /** Remove a post from bookmarks. */
  removeBookmark(params: {
    id: PostId;
    platform: Platform;
  }): Promise<results.Result<void, Error>>;
  /** List bookmarked posts for a platform. */
  getBookmarks(
    params: { platform: Platform; maxResults?: number },
  ): Promise<results.Result<Post[], Error>>;
}

/** Concrete use-case coordinator for multi-platform posting. */
export class PostService implements InboundPostService {
  private readonly apis: Map<Platform, SocialApi>;
  private readonly feedAggregator: FeedAggregator;
  private readonly translator: Translator;
  private readonly scheduler: Scheduler;

  constructor(
    connections: ReadonlyArray<PlatformConnection>,
    translator: Translator,
    scheduler: Scheduler,
    feedAggregator: FeedAggregator,
  ) {
    this.apis = new Map(connections.map((c) => [c.platform, c.socialApi]));
    this.feedAggregator = feedAggregator;
    this.translator = translator;
    this.scheduler = scheduler;
  }

  private resolveApiWithPlatformResult(
    platform?: Platform,
  ): results.Result<[Platform, SocialApi], Error> {
    if (platform !== undefined) {
      const api = this.apis.get(platform);
      if (api !== undefined) return results.ok([platform, api]);
      return results.fail(
        new Error(`Platform "${platform}" is not configured.`),
      );
    }
    for (const [p, api] of this.apis.entries()) {
      return results.ok([p, api]);
    }
    return results.fail(new Error("No platforms are configured."));
  }

  private resolveApiResult(
    platform?: Platform,
  ): results.Result<SocialApi, Error> {
    const r = this.resolveApiWithPlatformResult(platform);
    return results.map(r, ([, api]) => api);
  }

  composePost(
    text: string,
    platform?: Platform,
  ): Promise<results.Result<Post, Error>> {
    const resolveApiWithPlatformResult = this.resolveApiWithPlatformResult.bind(
      this,
    );
    return fn.run(async function* () {
      const [resolvedPlatform, api] = yield* resolveApiWithPlatformResult(
        platform,
      );
      const me = yield* (await results.fromPromise(api.getMe()));
      const { id, platformRef } = yield* (await results.fromPromise(
        api.createPost({ text }),
      ));
      return {
        id,
        text,
        authorHandle: me.handle,
        createdAt: new Date(),
        platform: resolvedPlatform,
        ...(platformRef !== undefined && { platformRef }),
      };
    });
  }

  async composePostToAll(
    text: string,
  ): Promise<results.Result<PostResult[], Error>> {
    const platforms = Array.from(this.apis.keys());
    const perPlatform = await Promise.all(
      platforms.map((p) => this.composePost(text, p)),
    );
    const postResults: PostResult[] = perPlatform.map((r, i) => {
      const platform = platforms[i] as Platform;
      return results.isOk(r)
        ? { platform, post: r.value }
        : { platform, error: r.error };
    });
    return results.ok(postResults);
  }

  translateAndPost(params: {
    text: string;
    from: string;
    to: string;
    platform?: Platform;
  }): Promise<results.Result<Post, Error>> {
    const translator = this.translator;
    const composePost = this.composePost.bind(this);
    return fn.run(async function* () {
      const translated = yield* (await translator.translate({
        text: params.text,
        from: params.from,
        to: params.to,
      }));
      return yield* (await composePost(translated, params.platform));
    });
  }

  schedulePost(
    params: { text: string; scheduledAt: Date; platform?: Platform },
  ): Promise<results.Result<void, Error>> {
    const scheduler = this.scheduler;
    return fn.run(async function* () {
      yield* (await results.fromPromise(
        scheduler.schedule({
          text: params.text,
          scheduledAt: params.scheduledAt,
        }),
      ));
    });
  }

  getUnifiedTimeline(
    params?: { maxResultsPerPlatform?: number },
  ): Promise<results.Result<Post[], Error>> {
    const feedAggregator = this.feedAggregator;
    return fn.run(async function* () {
      return yield* (await results.fromPromise(
        feedAggregator.getUnifiedTimeline(params),
      ));
    });
  }

  getTimeline(
    params?: { maxResults?: number; platform?: Platform },
  ): Promise<results.Result<Post[], Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params?.platform);
      return yield* (await results.fromPromise(
        api.getTimeline({ maxResults: params?.maxResults }),
      ));
    });
  }

  getPost(
    id: PostId,
    platform: Platform,
  ): Promise<results.Result<Post, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(platform);
      return yield* (await results.fromPromise(api.getPost(id)));
    });
  }

  replyToPost(
    params: { text: string; inReplyToPost: Post },
  ): Promise<results.Result<Post, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(
        params.inReplyToPost.platform,
      );
      const me = yield* (await results.fromPromise(api.getMe()));
      const { id, platformRef } = yield* (await results.fromPromise(
        api.replyToPost(params),
      ));
      return {
        id,
        text: params.text,
        authorHandle: me.handle,
        createdAt: new Date(),
        platform: params.inReplyToPost.platform,
        inReplyToId: params.inReplyToPost.id,
        ...(platformRef !== undefined && { platformRef }),
      };
    });
  }

  postThread(
    texts: ReadonlyArray<string>,
    platform?: Platform,
  ): Promise<results.Result<Post[], Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(platform);
      const result = yield* (await api.postThread({ texts }));
      return result.posts;
    });
  }

  translateAndReply(params: {
    text: string;
    inReplyToPost: Post;
    from: string;
    to: string;
  }): Promise<results.Result<Post, Error>> {
    const translator = this.translator;
    const replyToPost = this.replyToPost.bind(this);
    return fn.run(async function* () {
      const translated = yield* (await translator.translate({
        text: params.text,
        from: params.from,
        to: params.to,
      }));
      return yield* (await replyToPost({
        text: translated,
        inReplyToPost: params.inReplyToPost,
      }));
    });
  }

  getUsage(
    platform?: Platform,
  ): Promise<results.Result<UsageData, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(platform);
      return yield* (await results.fromPromise(api.getUsage()));
    });
  }

  repost(
    params: { id: PostId; platform: Platform },
  ): Promise<results.Result<void, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      yield* (await results.fromPromise(api.repost(params.id)));
    });
  }

  undoRepost(
    params: { id: PostId; platform: Platform },
  ): Promise<results.Result<void, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      yield* (await results.fromPromise(api.undoRepost(params.id)));
    });
  }

  quotePost(
    params: { text: string; quotedPostId: PostId; platform: Platform },
  ): Promise<results.Result<Post, Error>> {
    const resolveApiWithPlatformResult = this.resolveApiWithPlatformResult.bind(
      this,
    );
    return fn.run(async function* () {
      const [resolvedPlatform, api] = yield* resolveApiWithPlatformResult(
        params.platform,
      );
      const me = yield* (await results.fromPromise(api.getMe()));
      const { id } = yield* (await results.fromPromise(
        api.quotePost({
          text: params.text,
          quotedPostId: params.quotedPostId,
        }),
      ));
      return {
        id,
        text: params.text,
        authorHandle: me.handle,
        createdAt: new Date(),
        platform: resolvedPlatform,
        referencedPosts: [{ type: "quoted", id: params.quotedPostId }],
      };
    });
  }

  searchPosts(
    params: { query: string; maxResults?: number; platform: Platform },
  ): Promise<results.Result<Post[], Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      return yield* (await results.fromPromise(
        api.searchPosts({ query: params.query, maxResults: params.maxResults }),
      ));
    });
  }

  async searchPostsAll(
    params: { query: string; maxResultsPerPlatform?: number },
  ): Promise<results.Result<Post[], Error>> {
    const maxResults = params.maxResultsPerPlatform ?? 10;
    const platforms = Array.from(this.apis.keys());
    const perPlatform = await Promise.all(
      platforms.map((p) =>
        this.searchPosts({ query: params.query, maxResults, platform: p })
      ),
    );
    const all: Post[] = [];
    for (const r of perPlatform) {
      if (results.isOk(r)) all.push(...r.value);
    }
    return results.ok(
      all.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime()),
    );
  }

  bookmarkPost(
    params: { id: PostId; platform: Platform },
  ): Promise<results.Result<void, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      yield* (await results.fromPromise(api.bookmarkPost(params.id)));
    });
  }

  removeBookmark(
    params: { id: PostId; platform: Platform },
  ): Promise<results.Result<void, Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      yield* (await results.fromPromise(api.removeBookmark(params.id)));
    });
  }

  getBookmarks(
    params: { platform: Platform; maxResults?: number },
  ): Promise<results.Result<Post[], Error>> {
    const resolveApiResult = this.resolveApiResult.bind(this);
    return fn.run(async function* () {
      const api = yield* resolveApiResult(params.platform);
      return yield* (await results.fromPromise(
        api.getBookmarks({ maxResults: params.maxResults }),
      ));
    });
  }
}

File: pkg/ajan/postsfx/postservice.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package postsfx

import (
	"context"
	"fmt"
)

// DefaultPostService implements PostService using a registry of SocialApi adapters.
type DefaultPostService struct {
	reg *registry
}

// NewPostService creates a PostService backed by the given adapter registry.
func NewPostService(reg *registry) *DefaultPostService {
	return &DefaultPostService{reg: reg}
}

// ComposePost publishes a post to the specified platform (or first registered if nil).
func (s *DefaultPostService) ComposePost(ctx context.Context, opts ComposeOptions) (*Post, error) {
	api, err := s.resolveOne(opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.CreatePost(ctx, opts.Text)
}

// ComposePostToAll publishes to every registered platform.
func (s *DefaultPostService) ComposePostToAll(ctx context.Context, text string) ([]PostResult, error) {
	platforms := s.reg.Platforms()
	results := make([]PostResult, 0, len(platforms))

	for _, platform := range platforms {
		api, _ := s.reg.Get(platform)
		post, err := api.CreatePost(ctx, text)
		results = append(results, PostResult{Platform: platform, Post: post, Err: err})
	}

	return results, nil
}

// SchedulePost is a stub — scheduling requires persistence not yet implemented.
func (s *DefaultPostService) SchedulePost(_ context.Context, _ ScheduleOptions) error {
	return fmt.Errorf("SchedulePost: %w", ErrScheduleNotImpl)
}

// GetUnifiedTimeline merges timelines from all registered platforms.
func (s *DefaultPostService) GetUnifiedTimeline(ctx context.Context, maxResultsPerPlatform int) ([]*Post, error) {
	var all []*Post

	for _, platform := range s.reg.Platforms() {
		api, _ := s.reg.Get(platform)

		posts, err := api.GetTimeline(ctx, GetTimelineOptions{
			Platform:   platform,
			MaxResults: maxResultsPerPlatform,
		})
		if err != nil {
			continue // best-effort; skip failing platforms
		}

		all = append(all, posts...)
	}

	return all, nil
}

// GetTimeline fetches the timeline from a single platform.
func (s *DefaultPostService) GetTimeline(ctx context.Context, opts GetTimelineOptions) ([]*Post, error) {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.GetTimeline(ctx, opts)
}

// GetPost fetches a single post.
func (s *DefaultPostService) GetPost(ctx context.Context, id PostID, platform Platform) (*Post, error) {
	api, err := s.resolveOne(&platform)
	if err != nil {
		return nil, err
	}

	return api.GetPost(ctx, id)
}

// ReplyToPost publishes a reply.
func (s *DefaultPostService) ReplyToPost(ctx context.Context, opts ReplyOptions) (*Post, error) {
	if opts.InReplyToPost == nil {
		return nil, fmt.Errorf("ReplyToPost: %w", ErrReplyTargetRequired)
	}

	api, err := s.resolveOne(&opts.InReplyToPost.Platform)
	if err != nil {
		return nil, err
	}

	return api.ReplyToPost(ctx, opts)
}

// PostThread publishes a thread.
func (s *DefaultPostService) PostThread(ctx context.Context, opts ThreadOptions) ([]*Post, error) {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.PostThread(ctx, opts)
}

// GetUsage returns usage data for a single platform.
func (s *DefaultPostService) GetUsage(ctx context.Context, platform Platform) (*UsageData, error) {
	api, err := s.resolveOne(&platform)
	if err != nil {
		return nil, err
	}

	return api.GetUsage(ctx)
}

// Repost reposts.
func (s *DefaultPostService) Repost(ctx context.Context, opts RepostOptions) error {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return err
	}

	return api.Repost(ctx, opts)
}

// UndoRepost undoes a repost.
func (s *DefaultPostService) UndoRepost(ctx context.Context, opts RepostOptions) error {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return err
	}

	return api.UndoRepost(ctx, opts)
}

// QuotePost creates a quote post.
func (s *DefaultPostService) QuotePost(ctx context.Context, opts QuotePostOptions) (*Post, error) {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.QuotePost(ctx, opts)
}

// SearchPosts searches posts on a single platform.
func (s *DefaultPostService) SearchPosts(ctx context.Context, opts SearchOptions) ([]*Post, error) {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.SearchPosts(ctx, opts)
}

// SearchPostsAll searches across all registered platforms.
func (s *DefaultPostService) SearchPostsAll(ctx context.Context, query string, maxResultsPerPlatform int) ([]PostResult, error) {
	var results []PostResult

	for _, platform := range s.reg.Platforms() {
		api, _ := s.reg.Get(platform)

		posts, err := api.SearchPosts(ctx, SearchOptions{
			Query:      query,
			Platform:   platform,
			MaxResults: maxResultsPerPlatform,
		})

		for _, p := range posts {
			results = append(results, PostResult{Platform: platform, Post: p})
		}

		if err != nil {
			results = append(results, PostResult{Platform: platform, Err: err})
		}
	}

	return results, nil
}

// BookmarkPost bookmarks a post.
func (s *DefaultPostService) BookmarkPost(ctx context.Context, opts BookmarkOptions) error {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return err
	}

	return api.BookmarkPost(ctx, opts)
}

// RemoveBookmark removes a bookmark.
func (s *DefaultPostService) RemoveBookmark(ctx context.Context, opts BookmarkOptions) error {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return err
	}

	return api.RemoveBookmark(ctx, opts)
}

// GetBookmarks lists bookmarks.
func (s *DefaultPostService) GetBookmarks(ctx context.Context, opts GetBookmarksOptions) ([]*Post, error) {
	api, err := s.resolveOne(&opts.Platform)
	if err != nil {
		return nil, err
	}

	return api.GetBookmarks(ctx, opts)
}

// resolveOne returns the adapter for the given platform.
// If platform is nil or unknown, returns the first registered adapter.
func (s *DefaultPostService) resolveOne(platform *Platform) (SocialApi, error) {
	if platform != nil && *platform != PlatformUnknown && *platform != "" {
		api, ok := s.reg.Get(*platform)
		if !ok {
			return nil, fmt.Errorf("resolveOne %q: %w", *platform, ErrAdapterNotRegistered)
		}

		if api == nil {
			return nil, fmt.Errorf("resolveOne %q: %w", *platform, ErrNilAdapter)
		}

		return api, nil
	}

	platforms := s.reg.Platforms()
	if len(platforms) == 0 {
		return nil, fmt.Errorf("resolveOne: %w", ErrNoAdaptersRegistered)
	}

	api, _ := s.reg.Get(platforms[0])

	if api == nil {
		return nil, fmt.Errorf("resolveOne: %w", ErrNilAdapter)
	}

	return api, nil
}
